### ISSUE TITLE
Inside `FeedForward`, input goes through `dense1`, `dropout`, `dense2`. Is it intentional?

### ISSUE THREAD

**Issue by leoxzhao** [tensorflow/swift-models#104]
```swift
    func applied(to input: Tensor<Float>, in context: Context) -> Tensor<Float> {
        return input.sequenced(in: context, through: dense1, dropout, dense2)
    }
```
https://github.com/tensorflow/swift-models/blob/412a822a5c095c56504cae9681bd8ae1f8651edc/Transformer/Model.swift#L35

I did not find reference code in Python with dropout. I had the impression that dropout normally is added to the exit as the last operation before produces the final output. In this case, it would be dense1 -> dense2 -> dropout. This looks like a typo. But I don't possess the knowledge to judge. So I open an issue.

**Comment by jekbradbury** [comment]
Good catch!

The original Transformer performs dropout after the first dense layer in each feedforward block (see e.g. T2T code [here](https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/layers/common_layers.py#L1282-L1300)), but BERT performs it after the last dense layer in each feedforward block (see [code](https://github.com/google-research/bert/blob/master/modeling.py#L866-L879)).

This particular codebase is intended to replicate the architecture used in GPT-2, though, where we don't actually know which of these is the case since the training code isn't public.

**Comment by leoxzhao** [comment]
I re-read GPT-2 today. Compared with Python implementing of GPT-2, our EncoderLayer contains additional three dropout layers, `selfAttentionDropout`, `feedForwardDropout` and one inside of `FeedForward`.  Oddly, GPT-2 does not contain any dropout layers. I guess strategic placement of dropout layers one of the things that OpenAI is holding back. I am going to close this issue for now. We can revisit once we have more information or more development in this field.

**Comment by leoxzhao** [comment]
@xihui-wu I noticed that you recently did a fix for GPT2 training portion. If you are working on training or fine-tuning related aspect of it, this issue contains a conversation we had in the past about dropout which might be helpful.

HuggingFace's implementation also contains a dropout for embedding, https://github.com/huggingface/transformers/blob/2c1ebb8b507c19c75af5084b6c73e0b003c9eda6/src/transformers/modeling_gpt2.py#L483.

**Comment by xihui-wu** [comment]
@jekbradbury  >>The original Transformer performs dropout after the first dense layer in each feedforward block.    -- Looks like the link is outdated. Can you paste again?

@leoxzhao Yes as far as I see OpenAI don't show any dropout at all, whereas huggingface does. Compared with huggingface we have this extra dropout between two denses in MLP. I'd prefer to remove it to be aligned with huggingface whom we are taking as the baseline for benchmarking.

**Comment by jekbradbury** [comment]
I agree that you should align with huggingface.

**Comment by leoxzhao** [comment]
#574 has been merged. Closing this issue.
